src/typ.rs

```diff
@@ -2863,6 +2853,7 @@ pub fn infer_module(
 
                 // Infer the type
                 let mut fn_typer = typer.new_scope();
+                fn_typer.type_vars.clear();
```

Comment:
I needed to clear the `type_vars` here to get tests passing, but I'm not sure entirely sure why yet. I think type variables from custom type constructors may be leaking into the top level scope somewhere, but I haven't managed to confirm the issue + find a less hacky fix.

I'm going to keep looking, but I don't think this introduces any new bugs, because the original code used an empty `type_vars` hashmap here too. My main concern is whether there's a larger systemic problem for all module-level statements, but I haven't constructed any failing testcases yet.

Reply:
Makes sense, thank you. This seems like the same behaviour to me.